src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/ContrastRatioChecker.java

```diff
@@ -0,0 +1,177 @@
+        for (int i = 0; i < localMaximums.size(); i++) {
+            System.out.println(i + ": " + localMaximums.get(i).value + " --- " + localMaximums.get(i).totalOccurrence);
+        }
+
+        double [] contrastColors = get2MostPresentElements(localMaximums);
+        double colorContrast = getContrastRatio(contrastColors[0], contrastColors[1]);
+        return isContrastRatioCompliant(colorContrast, type, isHighVisibility);
+    }
+
+    public boolean isContrastRatioCompliant(double colorContrast, TextType type, boolean isHighVisibility) {
+        switch (type) {
+            case REGULAR:
+                return isHighVisibility ? colorContrast >= 7.0 : colorContrast >= 4.5;
+            case LARGE:
+                return isHighVisibility ? colorContrast >= 4.5 : colorContrast >= 3.0;
+            case LOGO:
+                return true;
+            default:
+                break;
+        }
+        return false;
+
+    }
+
+    private double relativeLuminosity(Color color) {
+        double normalizedRed = normalizeColorComponent(color.getRed());
+        double normalizeGreen = normalizeColorComponent(color.getGreen());
+        double normalizeBlue = normalizeColorComponent(color.getBlue());
+        return 0.2126 * normalizedRed + 0.7152 * normalizeGreen + 0.0722 * normalizeBlue;
+    }
+
+    private double normalizeColorComponent(int colorComponent) {
+        double doubleColorComponent = colorComponent/255.;
+        return doubleColorComponent < 0.03928 ? doubleColorComponent/12.92 :
+                Math.pow(((doubleColorComponent + 0.055)/1.055), 2.4);
+    }
+
+    private double getContrastRatio(double first, double second) {
+        double l1 = Math.max(first, second);
+        double l2 = Math.min(first, second);
+        return (l1 + 0.05) / (l2 + 0.05);
+    }
+
+    private List<DataPoint> getLuminosityPresenceList(BufferedImage bim) {
+
+        int width = bim.getWidth();
+        int height = bim.getHeight();
+        Map<Color, DataPoint> colorMap = new HashMap<>();
+
+        for (int i = 0; i < width; i++) {
+            for (int j = 0; j < height; j++) {
+                int rgb = bim.getRGB(i, j); //always returns TYPE_INT_ARGB
+                int alpha = (rgb >> 24) & 0xFF;// not used for now, should remove?
+                int red =   (rgb >> 16) & 0xFF;
+                int green = (rgb >>  8) & 0xFF;
+                int blue =  (rgb      ) & 0xFF;
+                Color color = new Color(red, green, blue);
+                double relativeLuminosity = relativeLuminosity(color);
+
+                if (colorMap.containsKey(color)) {
+                    colorMap.get(color).totalOccurrence++;
+                } else {
+                    colorMap.put(color, new DataPoint(relativeLuminosity));
+                }
+            }
+        }
+
+        return new ArrayList<>(new TreeSet<>(colorMap.values()));
+    }
+
+    private List<DataPoint> findLocalMaximums(List<DataPoint> source)
```

Comment:
findLocalMaximums can be implemented with 2 times less number of comparisons, just one comparison for each pair instead of 2 (we can store the "monotone" flag for previous pair and the peek local maximum if increase state is changed by decrease.

Reply:
ok, done